### DISCUSSION TITLE
How to add MarkupApp

### DISCUSSION THREAD

**Discussion by AntouanK** [iTwin/itwinjs-core#1006] [Q&A] [answered]
Hi there.

Got the web viewer demo app up and running ( https://www.itwinjs.org/learning/tutorials/develop-web-viewer/ ) and I'm trying now to add the Markup functionality ( https://www.itwinjs.org/reference/imodeljs-markup/markupapp/markup/?term=markup )

When trying to instantiate it, it asks for a `ScreenViewport`, and I'm not sure how I can provide one from the viewer I got so far.
Any help on how to do that?

Thank you.

**Comment by kabentley** [comment, marked as answer]
you can use `IModelApp.viewManager.selectedView` to get the viewport the user last used (the only viewport in this case.)

**Reply by AntouanK** [reply]
That react app guide doesn't show how to get an `IModelApp`.

I did something like that so far :

made a handler when the iModel is ready

```
  const onImodelReady = (iModel: IModelConnection) => {
    IModelApp.viewManager.onViewOpen.addOnce(async (vp) => {
      await MarkupApp.initialize();
      MarkupApp.start(vp);
    });
  };
```

and added it in a callback I found on my `<Viewer>` properties
```
  return (
    <div className="viewer-container">
      <Header
        handleLogin={onLoginClick}
        loggedIn={isAuthorized}
        handleLogout={onLogoutClick}
      />
      {isLoggingIn ? (
        <span>"Logging in...."</span>
      ) : (
        isAuthorized && (
          <Viewer
            contextId={process.env.IMJS_CONTEXT_ID ?? ""}
            iModelId={process.env.IMJS_IMODEL_ID ?? ""}
            authConfig={{ oidcClient: AuthorizationClient.oidcClient }}
            theme={"light"}
            onIModelConnected={onImodelReady}
          ></Viewer>
        )
      )}
    </div>
  );
```

That compiles fine.

What I get now is a yellow frame around the 3d view.
![2021-03-22_12-18](https://user-images.githubusercontent.com/4569111/111982041-c3729e00-8b08-11eb-802c-7e1b240985d9.png)


But I cannot click anything inside it anymore.
Wondering if that's what this feature do?

**Reply by kabentley** [reply]
You're already using `IModelApp` here:

 const onImodelReady = (iModel: IModelConnection) => {
    **IModelApp.viewManager**.onViewOpen.addOnce(async (vp) => {
      await MarkupApp.initialize();
      MarkupApp.start(vp);
    });
  };

you just added a listener to open Markup mode when the view is first opened (probably not what you want). The yellow border tells you that you're in markup mode. In markup mode you can use any of the `MarkupTool`s. Call `MarkupApp,stop()` to end markup mode and get the Svg data and image of the markup.

I think there's an extension that adds UI for markup mode @calebmshafer ?

**Reply by AntouanK** [reply]
Yeah, I wasn't sure if that's the right way to use `IModelApp` or not.

Ok, I suspected you need a toolbar to switch that mode on/off and get something out of it.
If you can point me to that, it would be great. Thank you.

**Reply by calebmshafer** [reply]
@AntouanK, we do have an Extension available [here](https://github.com/imodeljs/viewer-components-react/tree/master/packages/markup-frontstage).  This uses a full [frontstage](https://www.itwinjs.org/reference/ui-framework/frontstage/) to control when you're in Markup mode.

A bit smaller example is one I put together to start and stop the session using a toolbar buttons is available [here](https://github.com/calebmshafer/imodeljs-extension-demo/blob/main/my-extension/src/MyExtension.ts#L15-L61).

**Reply by AntouanK** [reply]
I'll give it a try.
Thank you. :+1:

**Reply by AntouanK** [reply]
@calebmshafer 
I tried adding an Extension exactly like the one you had in that example, and I get this error : 
![2021-03-23_10-01](https://user-images.githubusercontent.com/4569111/112120830-e5c6f300-8bbe-11eb-86eb-e647f523704b.png)

**Reply by AntouanK** [reply]
If I remove that last line in the file, it compiles.

Then in the `<Viewer>` I added it like that : 

```js

  const extensions = [new MyExtension("sample")];

  return (
    <div className="viewer-container">
      <Header
        handleLogin={onLoginClick}
        loggedIn={isAuthorized}
        handleLogout={onLogoutClick}
      />
      {isLoggingIn ? (
        <span>"Logging in...."</span>
      ) : (
        isAuthorized && (
          <Viewer
            contextId={process.env.IMJS_CONTEXT_ID ?? ""}
            iModelId={process.env.IMJS_IMODEL_ID ?? ""}
            authConfig={{ oidcClient: AuthorizationClient.oidcClient }}
            theme={"light"}
            onIModelConnected={onImodelReady}
            extensions={extensions}
          ></Viewer>
        )
      )}
    </div>
  );
```

but then when the app loads, I don't see it anywhere... How is it supposed to appear?

**Reply by calebmshafer** [reply]
The UiItemsProvider can be passed directly into the Viewer using prop `uiProviders` instead of using the `extensions` prop. Here's a small example I put together that is directly adding it to the iTwin Viewer. https://github.com/calebmshafer/itwin-viewer-markup-example/commit/e653fbab3fdb5eaa100401090b9e0531519e4861

**Reply by AntouanK** [reply]
oh, great! Thanks Caleb.

I tried it, and I get this error at runtime
![2021-03-23_15-37](https://user-images.githubusercontent.com/4569111/112163892-bda2b880-8bed-11eb-80f7-423d74e013f3.png)

**Reply by calebmshafer** [reply]
Do you still have `i18n` in the constructor?  In my example, I removed it since it wasn't explicitly user.

**Reply by AntouanK** [reply]
I've added it back, like that
```js
  public static i18n: I18N;

  public constructor(i18n: I18N) {
    MarkupToolbarProvider.i18n = i18n;
  }
```

but now when calling the MarkupToolbarProvider, it asks for an `I18N` argument.
not sure what to pass there.

if I remove it I get the error I posted above.

**Reply by AntouanK** [reply]
hey @calebmshafer 

I'm trying to deploy my app to a server we have, and I'm getting a CORS rejection at the `https://imsoidc.bentley.com/.well-known/openid-configuration` endpoint
Is there a way to fix that?

**Reply by calebmshafer** [reply]
hi @AntouanK, you will need to add the new URL, of the server you're deploying to, as a "redirect uri" on the [app registration page](https://developer.bentley.com/apps/).  Once this goes into effect, sometimes it can take a few minutes, you will no longer receive the CORS issue.

e.g. Add `https://mywebsite.mydomain.com/` to the list of "Redirect URIs"

**Reply by AntouanK** [reply]
I get an empty page at that url 
![2021-03-29_14-44](https://user-images.githubusercontent.com/4569111/112838360-4497db00-909d-11eb-9087-beeb82d96664.png)

**Reply by calebmshafer** [reply]
Sorry about that, not sure why it's not showing you any apps. @laurynasr  any idea?

In the meantime, can you use this url instead to access your apps, https://www.itwinjs.org/getting-started/registration-dashboard?tab=0?

**Reply by laurynasr** [reply]
~~You need to register an application on https://developer.bentley.com and use client_id of that application with the sample viewer. I think you are using the demo credentials that come with the sample.~~
I just noticed that the link @calebmshafer provided is incorrect -- it should be https://developer.bentley.com/my-apps

**Reply by AntouanK** [reply]
you are correct @laurynasr 
That is the right link.

But, it seems like I cannot save the new URIs.
As soon as I click "Save" an error pops up really fast ( cannot read it ) and it doesn't save anything.
Here's a video 
![Peek 2021-03-29 16-50](https://user-images.githubusercontent.com/4569111/112855271-f2f84c00-90ae-11eb-8945-4afff4aed8ae.gif)

**Reply by laurynasr** [reply]
I think the reason it's not working for you is because http protocol is only allowed for localhost; global URIs must use https.
The quick popup looks like a bug, we'll fix it.

**Reply by AntouanK** [reply]
@laurynasr 
You were right, I've enabled https and saved the URIs. 

But it still throws a CORS error.
The app is here if you want to have a look https://itwin-frontendvue.toc.systems/

**Reply by laurynasr** [reply]
Your application is still using http redirect_uri, please change that to match application configuration in developer.bentley.com
![image](https://user-images.githubusercontent.com/18733763/112946701-8331a200-913e-11eb-80a5-0b5f05aa2d93.png)
